### PR TITLE
Add Hexadecimal

### PR DESCRIPTION
**Pull request by SkeletalDemise**
Adds a hexadecimal decoder.

### CODE REVIEW

File: src/decoders/hexadecimal_decoder.rs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        results.update_checker(&checker_result);
+
+        results
+    }
+}
+
+/// Decodes hexadecimal to string
+fn hexadecimal_to_string(hex: &str) -> String {
+    // Remove all non-hexadecimal characters from the string
+    let hex = hex.replace(|c: char| !c.is_ascii_hexdigit(), "");
```
Comment (bee-san) [inline src/decoders/hexadecimal_decoder.rs:59]: If a string has non hex characters, does that mean it's not hex and we should skip it?

**Review by bee-san — CHANGES_REQUESTED** [review]

File: src/decoders/hexadecimal_decoder.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+    // Convert the hexadecimal string to a vector of bytes
+    let bytes = hex.as_bytes();
+
+    // Ensure that the vector of bytes has an even length, so it can be processed in pairs
+    if bytes.len() % 2 == 1 {
+        return String::new();
+    }
```
[the remark this answers is not in the document]
Comment (SkeletalDemise) [reply src/decoders/hexadecimal_decoder.rs:67]: Everything breaks when I do that, what's wrong with always returning a string? It's empty if it fails.

[the remark this answers is not in the document]
Comment (bee-san) [reply src/decoders/hexadecimal_decoder.rs:67]: > what's wrong with always returning a string? It's empty if it fails.

It is not the Rust way:tm: 

Change to this:

```rust
fn hexadecimal_to_string(hex: &str) -> Option<String> {
```

Return None, if it's okay return `Some(result)`

```rust
        trace!("Trying hexadecimal with text {:?}", text);
         let decoded_text = hexadecimal_to_string(text);
    if decoded_text.is_none() <-- fail the decoder
         let mut results = CrackResult::new(self, text.to_string());
         ```

[the remark this answers is not in the document]
Comment (SkeletalDemise) [reply src/decoders/hexadecimal_decoder.rs]: Done.

**Pull request merged**